libnmstate/validator.py

```diff
@@ -334,3 +341,61 @@ def _assert_ovs_lag_slave_count(iface_state):
                 raise NmstateOvsLagValueError(
                     f"OVS {ifname} LAG port {lag} has less than 2 slaves."
                 )
+
+
+def _validate_linux_bond(original_desired_state, current_state):
+    """
+    Raise NmstateValueError on these scenarios:
+        * Bond mode not defined.
+        * Original desire state contains illegal bond config.
+        * After merge, user's intention will cause illegal bong config.
+          For example: mac specified in desired_state without any bond options
+                       defined. While current state is fail_over_mac=1 with
+                       active-backup mode.
+    """
+    merged_desired_state = copy.deepcopy(original_desired_state)
+    merged_desired_state.merge_interfaces(current_state)
+
+    original_iface_states = {}
+    for iface_name, iface_state in original_desired_state.interfaces.items():
+        original_iface_states[iface_name] = iface_state
+
+    for iface_state in merged_desired_state.interfaces.values():
+        if iface_state[Interface.STATE] != InterfaceState.UP:
+            continue
+        if iface_state[Interface.TYPE] == InterfaceType.BOND:
+            if not iface_state.get(Bond.CONFIG_SUBTREE, {}).get(Bond.MODE):
+                raise NmstateValueError("Bond mode is not defined")
+
+            original_iface_state = original_iface_states.get(
+                iface_state[Interface.NAME], {}
+            )
+            _assert_bond_config_is_legal(iface_state, original_iface_state)
+
+
+def _assert_bond_config_is_legal(iface_state, original_iface_state):
+    """
+    When interface has mac defined, active-backup mode, fail_over_mac=1 after
+    merge, raise NmstateValueError, but don't raise except when mac does not
+    defined in original_iface_state.
+    """
+    mac = original_iface_state.get(Interface.MAC)
+
+    if not mac:
+        return
+
+    bond_config = iface_state[Bond.CONFIG_SUBTREE]
+    mode = bond_config[Bond.MODE]
+    fail_over_mac = bond_config.get(Bond.OPTIONS_SUBTREE, {}).get(
+        "fail_over_mac"
+    )
+
+    if (
+        mac
+        and mode == BondMode.ACTIVE_BACKUP
+        and fail_over_mac in ("1", 1, "active")
```

Comment:
There should really be a helper like `bond_option_enabled(FAIL_OVER_MAC)` or ideally in the future a property of the state class like `bond_state.option.fail_over_mac` which is True for active and False otherwise.

Reply:
I guess this kind of collides with the plan to normalize the bond option values that @EdDev is working on, so maybe it could be skipped.